sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+# -*- coding: utf-8 -*-
+"""Recommendation system generator.
+
+Generate recommendations based on TF-IDF representation and a KNN model.
+"""
+# Author: Arturo Amor
+# License: 3-clause BSD
+
+
+import numbers
+import re
+from collections import defaultdict
+from pathlib import Path
+
+from sphinx.errors import ConfigError
+from .backreferences import (
+    _thumbnail_div,
+    THUMBNAIL_PARENT_DIV,
+    THUMBNAIL_PARENT_DIV_CLOSE,
+)
+from .py_source_parser import split_code_and_text_blocks
+from .gen_rst import extract_intro_and_title
+from .utils import _replace_md5
+
+
+class ExampleRecommender:
+    """Compute content-based KNN-TF-IFD recommendation system.
+
+    Parameters
+    ----------
+    n_examples : int, default=5
+        Number of most relevant examples to display.
+    min_df : float in range [0.0, 1.0] or int, default=1
+        When building the vocabulary ignore terms that have a document frequency
+        strictly lower than the given threshold. If float, the parameter
+        represents a proportion of documents, integer represents absolute
+        counts. This value is also called cut-off in the literature.
+    max_df : float in range [0.0, 1.0] or int, default=1.0
+        When building the vocabulary ignore terms that have a document frequency
+        strictly higher than the given threshold. If float, the parameter
+        represents a proportion of documents, integer represents absolute
+        counts.
+
+    Attributes
+    ----------
+    file_names_ : list of str
+        The list of file names used for computing the similarity matrix.
+        The recommended examples are chosen among this list.
+
+    similarity_matrix_ : dense matrix
+        Fitted matrix of pairwise cosine similarities.
+    """
+
+    def __init__(self, *, n_examples=5, min_df=3, max_df=0.9):
+        self.n_examples = n_examples
+        self.min_df = min_df
+        self.max_df = max_df
+
+    def token_freqs(self, doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        return self.dict_freqs(token_generator)
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    def dict_vectorizer(self, data):
+        """Convert a dictionary of feature arrays into a matrix.
+
+        Parameters
+        ----------
+        data : list of dict
+            An iterable of dictionaries of feature arrays, where each key
+            corresponds to a feature name, and each value is an array of
+            feature values.
```

Comment:
I thought the dictionary values were a count of frequency (not array)?

Reply:
Comment addressed in 1189948412327cc4bd5979f47bb614215b062c92.